lnwallet/rpcwallet/sign_coordinator.go

```diff
@@ -0,0 +1,913 @@
+		// ignore the response.
+
+		select {
+		case <-s.quit:
+			return
+		default:
+		}
+	}
+}
+
+// WaitUntilConnected waits until the remote signer has connected. If the remote
+// signer does not connect within the configured connection timeout, an error is
+// returned.
+func (s *SignCoordinator) WaitUntilConnected() error {
+	return s.waitUntilConnectedWithTimeout(s.connectionTimeout)
+}
+
+// waitUntilConnectedWithTimeout waits until the remote signer has connected. If
+// the remote signer does not connect within the given timeout, an error is
+// returned.
+func (s *SignCoordinator) waitUntilConnectedWithTimeout(
+	timeout time.Duration) error {
+
+	select {
+	case <-s.clientConnected:
+		return nil
+
+	case <-s.quit:
+		return ErrShuttingDown
+
+	case <-time.After(timeout):
+		return ErrConnectTimeout
+
+	case <-s.doneReceiving:
+		return ErrNotConnected
+	}
+}
+
+// createResponseChannel creates a response channel for the given request ID and
+// inserts it into the responses map. The function returns a cleanup function
+// which removes the channel from the responses map, and the caller must ensure
+// that this cleanup function is executed once the thread that's waiting for
+// the response is done.
+func (s *SignCoordinator) createResponseChannel(requestID uint64) func() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	respChan := make(chan *signerResponse, 1)
+
+	// Insert the response channel into the map.
+	s.responses[requestID] = respChan
+
+	// Create a cleanup function that will delete the response channel.
+	return func() {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+
+		select {
+		case <-respChan:
+			// If we have timed out, there could be a very unlikely
+			// scenario where we did receive a response before we
+			// managed to grab the lock in the cleanup func.
+			// In that case, we'll just ignore the response.
+			// We should still clean up the response channel though.
+		default:
+		}
+
+		delete(s.responses, requestID)
+	}
+}
+
+// getResponse waits for a response with the given request ID and returns the
+// response if it is received. If the corresponding response from the remote
+// signer is a SignerError, the error message is returned. If the response is
+// not received within the given timeout, an error is returned.
+//
+// Note: Before calling this function, the caller must have created a response
+// channel for the request ID.
+func (s *SignCoordinator) getResponse(requestID uint64,
+	timeout time.Duration) (*signerResponse, error) {
+
+	s.mu.Lock()
+
+	// Verify that we have a response channel for the request ID.
+	if _, ok := s.responses[requestID]; !ok {
+		// It should be impossible to reach this case, as we create the
+		// response channel before sending the request.
+		s.mu.Unlock()
+
+		return nil, fmt.Errorf("no response channel found for "+
+			"request ID %d", requestID)
+	}
+
+	// Grab the response channel for the request ID.
+	respChan := s.responses[requestID]
```

Comment:
nit: just do `respChan, ok := s.responses[requestID]` before the `if` above, then directly use the value below?

Reply:
Thanks! This was changed in the push that edited this to use a SyncMap instead :)